### ISSUE TITLE
General issues with some syntaxes when using entities within variables

### ISSUE THREAD

**Issue by JustDylan23** [SkriptLang/Skript#1287]
```vb
on right click:
	if target living entity is set:
		myFunction(targeted entity)
		
function myFunction(e: entity):
	damage {_e} by 3 hearts
```
this should damage the targeted entity by 3 hearths, but it just doesn't do anything. it doesn't output any errors.

**Comment by Pikachu920** [comment]
Confirmed:
![image](https://user-images.githubusercontent.com/28607612/40935477-2764385c-67ed-11e8-88d4-a3716f64f57b.png)

**Comment by Blueyescat** [comment]
I think function paramaters has some issues, I have heard a few times issue about using has item condition with a player parameter. I will try that.

**Comment by Dylan123111** [comment]
```vb
on sneak toggle:
	player isn't sneaking
	myFunction(player)

function myFunction(p: player):
	broadcast "%{_p}%"
	damage {_p} by 3
```strange, it also doesn't work with players, though it broadcasts their names

**Comment by Dylan123111** [comment]
Another strange thing is that if you try to use play hurt on {_e} in the function it warns you that hurt can only be played on entities, this is useless in this case because variables are dynamic, the warn should only occure when using a invailid static type

**Comment by TheBentoBox** [comment]
Making this issue the main one for these types of bugs (the many we've referenced above in other issues). There's seemingly some general issues with how certain (or all?) expressions/effects, such as removing a potion effect or damaging an entity, work with entities stored in variables. Needs investigating.

**Comment by Nicofisi** [comment]
After some debugging - functions are not the problem in any of the mentioned issues, neither are entities. Variables are the problem.

---

**TL;DR** the issue described in OP seems to affect any syntax which contains an expression that accepts multiple types, and a variable of type other-than-the-first-one is used. As an example, the syntax of `EffHealth` is `damage %slots/livingentities/itemstack% ...`. When you do `damage {_some.slot} ...` it will work, but not `damage {_some.player} ... ` nor `damage {_some.itemstack} ...`

Let's keep using `EffHealth` as an example - it calls `damageables.getArray(e)`

https://github.com/SkriptLang/Skript/blob/c3febaedb89a7ed00223bcceb44971951f06f011/src/main/java/ch/njol/skript/effects/EffHealth.java#L108

`damageables` is a variable, and `Variable#getArray(Event)` (`Variable`s extend `Expression`) returns an empty array for entities. It is *supposed to* return:
* if the variable is a list variable, an array converted to the required type
* if the variable is a single variable, an array consisting only of one element, converted to the required type
* an empty array when the conversion or something else fails (because Skript always errors silently, *by an interesting design*)

The thing is that the parser somewhere (I'm yet to find the exact place in code, and update this issue with the respective information, unless someone does it before me) determines that the variable should be of the first possible type, a `Slot` in this case

![IntelliJ debugger output](https://vgy.me/oi57El.png)

Since a `CraftPlayer` cannot be converted to a `Slot`, the returned array is empty, hence nothing is looped in `EffHealth`, and as a result nothing happens.

---

- #1313 turned out to be a very different issue
- #1363 - I didn't manage to reproduce it
- #1393 is an exact duplicate of the bug in OP
- #1324 - I don't see how this could be related

**Comment by bensku** [comment]
@Nicofisi Sorry for late response. Anyway... The parser bug I explained in that comment was supposedly fixed in dev30. I'm a bit puzzled now...

**Comment by noftaly** [comment]
Hi,
I've found this bug, wich is related to this issue : 
```vb
command /cmd1:
    trigger:
        while player is online:
            broadcast "&a1 ok"
            wait 2 seconds

function funcWhile(p: offline player):
    while {_p} is online:
        broadcast "&a2 ok"
        wait 2 seconds
command /cmd2:
    trigger:
        funcWhile(player)
```

/cmd1 works but not /cmd2 (there is not broadcast)

**Comment by Nicofisi** [comment]
@noftaly "completed" means that this will be available in the next Skript release :)

**Comment by noftaly** [comment]
Oh ok ty 😃
